kythe/go/serving/pipeline/beam.go

```diff
@@ -58,6 +75,162 @@ func FromEntries(s beam.Scope, entries beam.PCollection) *KytheBeam {
+	return accum
+}
+
+func (c *combineDecorPieces) AddInput(accum *srvpb.FileDecorations, p *ppb.DecorationPiece) *srvpb.FileDecorations {
+	if ref := p.GetReference(); ref != nil {
+		var kind string
+		if k := ref.GetKytheKind(); k == scpb.EdgeKind_UNKNOWN_EDGE_KIND {
+			kind = ref.GetGenericKind()
+		} else {
+			kind = schema.EdgeKindString(k)
+		}
+		accum.Decoration = append(accum.Decoration, &srvpb.FileDecorations_Decoration{
+			Anchor: &srvpb.RawAnchor{
+				StartOffset: ref.Anchor.Span.Start.ByteOffset,
+				EndOffset:   ref.Anchor.Span.End.ByteOffset,
+			},
+			Kind:   kind,
+			Target: kytheuri.ToString(ref.Source),
+		})
+	} else if file := p.GetFile(); file != nil {
+		accum.File = file
+	} else if node := p.GetNode(); node != nil {
+		n := &srvpb.Node{Ticket: kytheuri.ToString(node.Source)}
+		if kind := nodes.Kind(node); kind != "" {
+			n.Fact = append(n.Fact, &cpb.Fact{
+				Name:  facts.NodeKind,
+				Value: []byte(kind),
+			})
+		}
+		if subkind := nodes.Subkind(node); subkind != "" {
+			n.Fact = append(n.Fact, &cpb.Fact{
+				Name:  facts.Subkind,
+				Value: []byte(subkind),
+			})
+		}
+		for _, f := range node.Fact {
+			n.Fact = append(n.Fact, &cpb.Fact{
+				Name:  nodes.FactName(f),
+				Value: f.Value,
+			})
+		}
+		sort.Slice(n.Fact, func(i, j int) bool { return n.Fact[i].Name < n.Fact[j].Name })
+		accum.Target = append(accum.Target, n)
+	}
+	return accum
+}
+
+func (c *combineDecorPieces) ExtractOutput(fd *srvpb.FileDecorations) *srvpb.FileDecorations {
+	sort.Slice(fd.Decoration, func(i, j int) bool {
+		if c := compare.Ints(int(fd.Decoration[i].Anchor.StartOffset), int(fd.Decoration[j].Anchor.StartOffset)); c != compare.EQ {
+			return c == compare.LT
+		} else if c := compare.Ints(int(fd.Decoration[i].Anchor.EndOffset), int(fd.Decoration[j].Anchor.EndOffset)); c != compare.EQ {
+			return c == compare.LT
+		} else if c := compare.Strings(fd.Decoration[i].Kind, fd.Decoration[j].Kind); c != compare.EQ {
+			return c == compare.LT
+		}
+		return fd.Decoration[i].Target < fd.Decoration[j].Target
+	})
+	sort.Slice(fd.Target, func(i, j int) bool { return fd.Target[i].Ticket < fd.Target[j].Ticket })
+	return fd
+}
+
+func fileToDecorPiece(src *spb.VName, f *srvpb.File) (*spb.VName, *ppb.DecorationPiece) {
+	return src, &ppb.DecorationPiece{Piece: &ppb.DecorationPiece_File{f}}
+}
+
+func refToDecorPiece(r *ppb.Reference) (*spb.VName, *ppb.DecorationPiece, error) {
+	_, file, err := toEnclosingFile(r)
+	if err != nil {
+		return nil, nil, err
+	}
+	return file, &ppb.DecorationPiece{
+		Piece: &ppb.DecorationPiece_Reference{&ppb.Reference{
+			Source: r.Source,
+			Kind:   r.Kind,
+			Anchor: r.Anchor,
+		}},
+	}, nil
+}
+
+func fileVName(anchor *spb.VName) *spb.VName {
+	return &spb.VName{
+		Corpus: anchor.Corpus,
+		Root:   anchor.Root,
+		Path:   anchor.Path,
+	}
+}
+
+func nodeToDecorPiece(key *spb.VName, node func(**ppb.Node) bool, file func(**spb.VName) bool, emit func(*spb.VName, *ppb.DecorationPiece)) {
```

Comment:
Can you help me understand how Beam provides the *ToDecorPiece functions with their inputs?  (I assume you're doing so correctly, but I don't understand the mechanism at all.)

Reply:
I'll take `nodeToDecorPiece` as an example.

`nodeToDecorPiece` is a ParDo function called on line 89 and takes a CoGBK as its input.  CoGBKs can be thought as a set of tuples with the first component being a key and then each subsequent component being a collection of values (a GBK is essentially just a CoGBK of 1 collection of values per key).  Beam represents the collections as streams/iterators (i.e. that's `func(*ppb.Node) bool` and `func(**spb.VName) bool`).  The key is just the `*spb.VName`.

The CoGBK values come from the input tables to the CoGBK on line 90: the PCollection of `KV<*spb.VName, *ppb.Node>` and the PColleciton of `KV<*spb.VName, *srvpb.File>`.  Each table is grouped by key and its value collection per key becomes the stream passed to the subsequent ParDo.  Each tuple is passed independently to `nodeToDecorPiece` to be processed.